### PR TITLE
Windows support

### PR DESCRIPTION
**Pull request by BartKnucle**
Update from the previous changes using array join function instead of string concatenation.

### CODE REVIEW

**Comment by BartKnucle** [comment]
Replace the PR #595

**Comment by BartKnucle** [comment]
Can someone run the checks? I do not have sufficients rights...

**Comment by BartKnucle** [comment]
I created a "orbit-db-win" npm package to be able to test without manaing forks https://www.npmjs.com/package/orbit-db-win

**Comment by aphelionz** [comment]
@BartKnucle Very glad to be able to revisit this now that CircleCI seems to have windows support. Sorry that it took us so long to circle back.

**Comment by BartKnucle** [comment]
@aphelionz , I merge the upstream and still no luck... (But browsers tests?)

**Comment by aphelionz** [comment]
Thanks @BartKnucle! We'll have a look

**Comment by BartKnucle** [comment]
@aphelionz I add a look at path.posix.join when I discovered that the path.join did not output the same paths on linux and windows but the function just do not exist :)
I just commit some few changes. The "cp" command in package.json do not work on windows. So I hard link the librairies in the browser html directly to the concerned librairies. More tests passed 💃

File: src/orbit-db-address.js [hunk 1]
```diff
@@ -11,7 +10,7 @@ class OrbitDBAddress {
   }
 
   toString () {
-    return path.join('/orbitdb', this.root, this.path)
+    return ['/orbitdb', this.root, this.path].join('/')
```
[the remark this answers is not in the document]
Comment (BartKnucle) [reply src/orbit-db-address.js]: I have tried before that function with no luck...

const path = require('path')
console.log(path.posix.join('c:\\Mypath\\subdir\\', '/test'))

return: c:\Mypath\subdir\/test

**Comment by BartKnucle** [comment]
@aphelionz , @haadcode  I think we are now close. I still get 2 errors on the tests but I stuck... Maybe you can help?
Regards

**Comment by BartKnucle** [comment]
Still looking for a the answer. I think that maybe the old databases and cache located in the fixtures directory are note compatible with windows... Do you have the code you used to generate those databases? I gonne try to generate them again under Windows and rerun the tests.

**Comment by aphelionz** [comment]
Thanks @BartKnucle. I'm still not in love with the `[].join("/")` solution but if that's what we have to do that's what we have to do! Did you see `test/browser.spec.js`? That might be useful in seeing how we load in fixture data for other environments.

**Comment by mbchandru** [comment]
I run OrbitDb on WIndows 7, 64 bit, and get an error

E:\orbitDbFarm>orbitdb create farm feed
Error: Not a valid OrbitDB address: \orbitdb\zdpuAyYx6ukDL44xgc4hsEAcBdt5i2jbWkp
3BZdeR3zzoQDhB\farm

Please help

**Comment by BartKnucle** [comment]
> 
> 
> Thanks @BartKnucle. I'm still not in love with the `[].join("/")` solution but if that's what we have to do that's what we have to do! Did you see `test/browser.spec.js`? That might be useful in seeing how we load in fixture data for other environments.

@aphelionz I agree with you, the [].join('/') is not well suited for the file path construction, And I think we can do better for the directory option and database location. But for the database name construction, the result must be the same on all OS. I may not have the knowledge to go further in the investigations...

**Comment by BartKnucle** [comment]
> 
> 
> I run OrbitDb on Windows 7, 64 bit, and get an error
> 
> E:\orbitDbFarm>orbitdb create farm feed
> **Error: Not a valid OrbitDB address:** \orbitdb\zdpuAyYx6ukDL44xgc4hsEAcBdt5i2jbWkp
> 3BZdeR3zzoQDhB\farm
> 
> Please help

@mbchandru you have 2 options if you wanna use orbit under windows: Either use the 'windows' fork source code that follow the current codebase. Or use this librairy https://www.npmjs.com/package/orbit-db-win but It may be very buggy...

**Comment by mbchandru** [comment]
> > I run OrbitDb on Windows 7, 64 bit, and get an error
> > E:\orbitDbFarm>orbitdb create farm feed
> > **Error: Not a valid OrbitDB address:** \orbitdb\zdpuAyYx6ukDL44xgc4hsEAcBdt5i2jbWkp
> > 3BZdeR3zzoQDhB\farm
> > Please help
> 
> @mbchandru you have 2 options if you wanna use orbit under windows: Either use the 'windows' fork source code that follow the current codebase. Or use this librairy https://www.npmjs.com/package/orbit-db-win but It may be very buggy...

Thanks for the reply.

**Comment by mbchandru** [comment]
> > > I run OrbitDb on Windows 7, 64 bit, and get an error
> > > E:\orbitDbFarm>orbitdb create farm feed
> > > **Error: Not a valid OrbitDB address:** \orbitdb\zdpuAyYx6ukDL44xgc4hsEAcBdt5i2jbWkp
> > > 3BZdeR3zzoQDhB\farm
> > > Please help
> > 
> > 
> > @mbchandru you have 2 options if you wanna use orbit under windows: Either use the 'windows' fork source code that follow the current codebase. Or use this librairy https://www.npmjs.com/package/orbit-db-win but It may be very buggy...
> 
> Thanks for the reply.

I tried by downloading the windows fork. Did npm install in that directory. Before that I uninstalled old orbit db and its cli from npm.
Same error
E:\>orbitdb create orbitDbFarm feed
Error: Not a valid OrbitDB address: \orbitdb\zdpuAyD4uTMavSpzECgwH7kHYtj4BkqA5FQ
mVY8vh3QQBxDCh\orbitDbFarm

**Comment by aphelionz** [comment]
We were discussing this and I think @shamb0t found a good place in the code where you might be able to abstract away all the `[].join('/')` calls to a single place, specifically here: https://github.com/orbitdb/orbit-db/blob/94c99d458fccfb950dcde9b9080a0a1f1004f185/src/orbit-db-address.js#L14

Our thinking is that while `[].join('/')` might not be a good pattern to use over and over again, if we can contain it to the `toString` function, we can achieve Windows compatibility with minimal changes. It may require replacing some things with `OrbitDBAddress.toString` but at least we can later refactor this singularly if we find a better approach.

What do you think @BartKnucle?

**Comment by BartKnucle** [comment]
@aphelionz, @shamb0t I'll have to give up on this. I tried to find an elegant solution to handle the "path" problem but since the problem is that windows introduce backslash everywhere you use the path.join(), everything is broken. Windows is able to handle the poisix paths. So for removing the use of the 'path' library is the way to go... Sorry guys. And for the errors that persists for the cache loading, I had a look at 'test/browser.spec.js' and didn't found something relevant... :(

**Comment by RichardLitt** [comment]
Thanks for your work, @BartKnucle. Jump on the community call in a couple of Tuesdays. :) We'd be happy to talk about it there.

**Comment by csuwildcat** [comment]
Any word on Windows support? Our team was looking into using Orbit, but without the cross platform support, we're unable to proceed.

**Comment by RichardLitt** [comment]
@csuwildcat Windows is currently not supported. If you would like to help with this effort and take up the cause of maintaining Windows support, you're welcome to, though!

**Comment by dacom-dark-sun** [comment]
Hello, Guys!. Great project, and I already created a prototype of the file storage as a dropbox and an encrypted editor with access levels. And oops! The lack of support for Windows, I somehow overlooked ...: ((

I decided to test in a browser, made a deploy to the server and checked it on a computer with windows - everything works. I believe that everything will works in the Electron shell too. Thus, there is no only native support for Windows, but everything works in browser or electron. I'm right?

**Comment by mbchandru** [comment]
I tried and failed on windows. Moved on to Ubuntu, which had
        its own problems.
      Lastly ended up using orbit-http-api from github, which has its
        own orbit-db sub install.
      I had not tested this on windows.
      
    
    Hope this helps.
    
      Regards,
Chandramoulee

**Comment by dacom-dark-sun** [comment]
I'we build electron app for windows, so, it not works. %( The package for Windows is temporarily canceled, it seems .. Unfortunately, I do not have a working environment in order to help with solving the problem: ((

**Comment by csuwildcat** [comment]
I have it running in WSL now, but it's a harder sell to get our folks on
board with supporting a project for the long haul when this sort of stuff
isn't a priority for the team leading the project. I work for Microsoft,
who naturally has a strong preference for code/projects that are friendly
for developers to run across major envs. (btw, I'm not saying any of this
to be negative, it's just advice - as you know, we learned a few humbling
lessons ourselves about the importance of this)

**Comment by kolessios** [comment]
Right now I am working on a project with OrbitDB and like everyone here I need Windows support so using the information that has been collected in this PR it seems to me that I have been able to make orbit-db work in Windows. Actually it's just changing some few lines so it's possible that I'm doing it wrong and it doesn't work on all versions of Windows but so far all test commands work for me in Windows 10 1903 (`examples:node`, `examples:browser-windows`, `build`, `test:browser-multiple-tabs`)



The only thing that stops me from doing a new PR is one of the problems mentioned here: there are 4 tests that fail only in Windows.



From what I see is the test to load the cache of a previous version and I have no idea how to fix it.

If anyone is interested, this is the repository with my changes:
https://github.com/dreamnettech/orbit-db/tree/feature/windows-support

**Comment by RichardLitt** [comment]
Thanks for the feedback, @csuwildcat and others. It's a tough call. None of the core maintainers use windows - so this doesn't necessarily scratch our itch. Right now, it's not a priority, but that doesn't mean that we're not 100% totally open to someone else taking over keeping Windows availability running. If you're up for being the point person for Windows, do step up! 

> From what I see is the test to load the cache of a previous version and I have no idea how to fix it.

If this is the only issue barring us from merging this PR and others like it, it might be worth excluding those tests (since they seem to load a previous version, which by its nature would be broken). @aphelionz Does those sound right to you?

**Comment by kolessios** [comment]
@RichardLitt I have already fixed that problem :) #712

**Comment by RichardLitt** [comment]
Closing now that https://github.com/orbitdb/orbit-db/pull/712 has been merged. Thank you all for your help! Especially you, @BartKnucle. :)

**Comment by BartKnucle** [comment]
@RichardLitt a way better work from @kolessios 👍 This is just what we needed. Happy that windows support finally appended!!! Thx all!

**Pull request closed without merging**